File: backend/app/schemas/datasets.py
# Response schemas for /api/datasets endpoint; wire format decoupled from SQLModel tables.

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class DatasetOut(BaseModel):
    """One dataset row as returned to API clients."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    display_name: str
    citation: str
    license: str
    cell_count: int
    gene_count: int
    created_at: datetime


class DatasetsResponse(BaseModel):
    """Full list response for GET /api/datasets."""

    items: list[DatasetOut]
    total: int

File: frontend/lib/schemas/datasets.ts
// Zod schemas for /api/datasets response — must stay in sync with backend/app/schemas/datasets.py.

import { z } from "zod";

export const DatasetSchema = z.object({
  id: z.string().uuid(),
  slug: z.string().min(1),
  display_name: z.string().min(1),
  citation: z.string(),
  license: z.string(),
  cell_count: z.number().int().nonnegative(),
  gene_count: z.number().int().nonnegative(),
  created_at: z.string().datetime(),
});

export const DatasetsResponseSchema = z.object({
  items: z.array(DatasetSchema),
  total: z.number().int().nonnegative(),
});

export type Dataset = z.infer<typeof DatasetSchema>;
export type DatasetsResponse = z.infer<typeof DatasetsResponseSchema>;
